### PR TITLE
Add support for Managed Service for Prometheus

### PR DESCRIPTION
**Pull request by mbarrien**
Adds support for Managed Service for Prometheus to google_container_cluster. Closes https://github.com/hashicorp/terraform-provider-google/issues/11224

This corresponds to https://github.com/hashicorp/terraform-provider-google-beta/pull/4105 where I was told to make the PR here.

Configuration looks like the following:
```
resource "google_container_cluster" {
  monitoring_config {
    managed_prometheus {
      enabled = true
    }
  }
}
```

Open Issues:
* To do this, makes monitoring_config.enabled_components into an optional, to allow configuring just the managed Prometheus component. Without further tests (that might require me to spin up an actual cluster), it's unclear if leaving that empty when enabling just the Prometheus component does the right thing.
* Unclear whether I need to add a ForceSendFields = ["Enabled"] to ensure that prometheus is properly disabled if the component is removed.
* Unable to run the make test and make lint on generated code; was able to generate the code, but no makefile exists in the generated code to run make test or make lint in.
* Unable to run acceptance tests myself; not set up for it and lack an account with the budget to do it in.

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
container: add `managed_prometheus` to `monitoring_config` in `google_container_cluster` (beta)
```

### CODE REVIEW

File: mmv1/third_party/terraform/resources/resource_container_cluster.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/resources/resource_container_cluster.go.erb:3671]: It seems like to make this simpler, the managed_prometheus handling could be the only thing behind the `<% if version == 'beta' %>`.  I think that would leave the behavior of the ComponentConfig the same, since it should always be specified in GA.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4981]: Since one of the changes is to make this optional - would it be possible to remove it as a field in this config?

**Review by melinath — COMMENTED** [review]
Generally LGTM assuming the vcr tests pass - a couple of minor thoughts in the meantime.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb
Comment (mbarrien) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4981]: Already implicitly done at line 1895 when we restore to testAccContainerCluster_basic at the end of  TestAccContainerCluster_withMonitoringConfig.

File: mmv1/third_party/terraform/resources/resource_container_cluster.go.erb
Comment (mbarrien) [reply mmv1/third_party/terraform/resources/resource_container_cluster.go.erb:3663]: Done.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb
Comment (melinath) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4981]: That tests that monitoring_config as a whole is optional; having enable_components be unset in this test would ensure that it is optional even if a monitoring_config is provided. This is not necessarily a blocker, but it would be nice unless there's a particular reason this needs to be set.

Comment (mbarrien) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4998]: Done.

**Comment by shuyama1** [comment]
FYI @melinath Looks like the provider crashes while running the VCR tests. Please ignore the generated comment on new VCR test result (fixing them now). Sorry for the confusion!

**Comment by shuyama1** [comment]
Fixed VCR test script. Testing it now.
/gcbrun

**Comment by melinath** [comment]
@shuyama1 the failure doesn't seem to be related - just want to double-check that the pipeline is fixed now?

**Comment by shuyama1** [comment]
I think the code added in this PR makes the provider crashed (`Test ended in panic` may be the more accurate wording, sorry for the confusion).
Looks like it also happens in TC, such as https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstreamVcr/275971?buildTab=tests&status=failed&expandedTest=-5350428937263695653

**Comment by mbarrien** [comment]
It's going to be difficult for me to help since I lack access to the logs, as a non-Googler (long long ago ex-Googler). Can anyone copy and paste?

If this repo provided the ability for maintainer to push into my branch, I believe I would've given those too.

**Review by melinath — CHANGES_REQUESTED** [review]
@mbarrien okay so the error is:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x27912cf]

goroutine 140229 [running]:
github.com/hashicorp/terraform-provider-google-beta/google-beta.flattenMonitoringConfig(...)
	/go/src/github.com/hashicorp/terraform-provider-google-beta/google-beta/resource_container_cluster.go:4070
github.com/hashicorp/terraform-provider-google-beta/google-beta.resourceContainerClusterRead(0xc003680f00, 0x3a22ac0, 0xc012c10000, 0x33dc660, 0x55cfa20)
	/go/src/github.com/hashicorp/terraform-provider-google-beta/google-beta/resource_container_cluster.go:1923 +0x358f
```

Here's the relevant line: https://github.com/modular-magician/terraform-provider-google-beta/blob/60622501e27affcdbb6272bd19d41b4bff45da75/google-beta/resource_container_cluster.go#L4070

I think probably c.ManagedPrometheusConfig is a nil pointer so accessing .Enabled on it panics. Are you able to run the tests locally and reproduce this?

**Comment by mbarrien** [comment]
Pushed a fix to flattenMonitoringConfig to make both of those fields truly treated as optional. Visually verified that expandMonitoringConfig is already doing the right thing with respect to those fields being optional. Please re-review/retest.

**Review by melinath — CHANGES_REQUESTED** [review]
The downstream is currently failing to compile:

```
/go/pkg/mod/github.com/modular-magician/terraform-provider-google@v1.20.1-0.20220406180705-23fac4291df7/google/resource_container_cluster.go:3444:9: non-integer slice index "enable_components"
/go/pkg/mod/github.com/modular-magician/terraform-provider-google@v1.20.1-0.20220406180705-23fac4291df7/google/resource_container_cluster.go:3444:31: cannot use c.ComponentConfig.EnableComponents (type []string) as type map[string]interface {} in assignment
```

**Comment by mbarrien** [comment]
Dang it, didn't copy those fixes over after confirming it compiled in my standalone terraform-provider-google-beta. Pushed fixes now.

**Comment by mbarrien** [comment]
Should be ready for retest.

**Comment by melinath** [comment]
@mbarrien I'm rerunning the tests but I don't see any new changes in this PR - just want to double-check that everything is pushed?

**Comment by mbarrien** [comment]
Grr.... having issues with multiple git remotes. Try now. So sorry for the noise!

**Comment by melinath** [comment]
It looks like the compilation failure should be resolved at this point - I'm not sure what happened but I'm gonna give it another go /gcbrun

File: mmv1/third_party/terraform/resources/resource_container_cluster.go.erb
Comment (mbarrien) [reply mmv1/third_party/terraform/resources/resource_container_cluster.go.erb]: Removal done. I think when I wrote that I was thinking we were one level up in the tree, and not in the actual list of component names.

File: mmv1/third_party/terraform/resources/resource_container_cluster.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/resources/resource_container_cluster.go.erb:3679]: This seems to be causing a panic:

```
panic: interface conversion: interface {} is string, not []interface {}

goroutine 2274 [running]:
github.com/hashicorp/terraform-provider-google-beta/google-beta.expandMonitoringConfig(0x33dac00, 0xc002037ea8, 0x11)
	/go/src/github.com/hashicorp/terraform-provider-google-beta/google-beta/resource_container_cluster.go:3592 +0x345
```

Downstream diff link: https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5812-old..auto-pr-5812#diff-c173cfcd7737c1477299c920117b36a0aa5570f842b93f5e07ccb383b4157947R3592

It looks like the logic changed slightly compared to the old version? I'm guessing that restoring the old conversion will fix this.

```suggestion
		enable_components := v.([]interface{})
```

**Review by melinath — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/resources/resource_container_cluster.go.erb
Comment (pintohutch) [reply mmv1/third_party/terraform/resources/resource_container_cluster.go.erb:3679]: @mbarrien - are you still working on this feature?

I wanted to call out that the feature has been promoted from beta to [GA](https://pkg.go.dev/google.golang.org/api/container/v1#MonitoringConfig) on the GKE API. So you shouldn't need to guard the API around `beta` anymore.

**Comment by mbarrien** [comment]
Pushed some changes to fix the identified bug, and to remove the beta guards. Please re-test.

**Comment by melinath** [comment]
`TestAccServiceNetworkingPeeredDNSDomain_basic` failure is unrelated. BillingSubaccount tests seem like flakes - re-running here: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/294874

**Review by melinath — CHANGES_REQUESTED** [review]
It looks like this change fails in the GA provider with:

```
Error: expected monitoring_config.0.enable_components.1 to be one of [SYSTEM_COMPONENTS], got WORKLOADS
on terraform_plugin_test.tf line 7, in resource "google_container_cluster" "primary":
7:          enable_components = [ "SYSTEM_COMPONENTS", "WORKLOADS" ]
```

It looks like this is due to WORKLOADS only being available in beta. If that's still accurate, then the part of the test that adds WORKLOADS needs to be gated so it is only present in beta.

**Comment by melinath** [comment]
Closed - changes were merged in https://github.com/GoogleCloudPlatform/magic-modules/pull/6104

**Pull request closed without merging**
